lib/Sema/CSDiag.cpp

```diff
@@ -5317,6 +5317,96 @@ static bool diagnoseSingleCandidateFailures(CalleeCandidateInfo &CCI,
       .diagnose();
 }
 
+static bool isRawRepresentableMismatch(Type fromType, Type toType,
+                                       KnownProtocolKind kind,
+                                       const ConstraintSystem *CS) {
+  toType = toType->lookThroughAllAnyOptionalTypes();
+  fromType = fromType->lookThroughAllAnyOptionalTypes();
+
+  // First check if this is an attempt to convert from something to
+  // raw representable.
+  if (conformsToKnownProtocol(fromType, kind, CS)) {
+    if (auto rawType = isRawRepresentable(toType, kind, CS))
+      return true;
+  }
+
+  // Otherwise, it might be an attempt to convert from raw representable
+  // to its raw value.
+  if (auto rawType = isRawRepresentable(fromType, kind, CS)) {
+    if (conformsToKnownProtocol(toType, kind, CS))
+      return true;
+  }
+
+  return false;
+}
+
+static bool diagnoseRawRepresentableMismatch(CalleeCandidateInfo &CCI,
+                                             Expr *argExpr,
+                                             ArrayRef<Identifier> argLabels) {
+  // We are only interested in cases which are
+  // unrelated to argument count or label mismatches.
+  switch (CCI.closeness) {
+    case CC_OneArgumentNearMismatch:
+    case CC_OneArgumentMismatch:
+    case CC_OneGenericArgumentNearMismatch:
+    case CC_OneGenericArgumentMismatch:
+    case CC_ArgumentNearMismatch:
+    case CC_ArgumentMismatch:
+      break;
+
+    default:
+      return false;
+  }
+
+  auto argType = argExpr->getType();
```

Comment:
Can you use `ConstraintSystem::getType(argExpr)` here?

I'm inching closer to getting all the CS code switched over to the side map but new appearances of `Expr::getType()` keep creeping in.